model/pipeline/src/main/resources/org/apache/beam/model/pipeline/pipeline.json

```diff
@@ -0,0 +1,320 @@
+{
+  "$schema": "http://json-schema.org/draft-04/schema#",
+
+  "type": "object",
+
+  "$ref": "#/definitions/Pipeline",
+
+  "definitions": {
+    // A Pipeline contains the entire hierarchical graph including
+    // nodes for all values.
+    //
+    // Nodes (transform or value or other) are identified by some unique id. This need
+    // not be readable. It is simplest to think of it as a reified pointer or UUID.
+    // Cyclic pointer structures are forbidden throughout.
+    "Pipeline": {
+      "type": "object",
+      "additionalProperties": false,
+      "required": [ "userFns", "transforms", "values", "windowingStrategies", "coders" ],
+      "properties": {
+
+        // A mapping containing all user-definable functions in the pipeline.
+        // The keys must be unique to this pipeline definition, but otherwise are
+        // arbitrary.
+        "userFns": {
+          "type": "object",
+          "additionalProperties": { "$ref": "#/definitions/UserFn" }
+        },
+
+        // A mapping containing all transforms in the pipeline.
+        "transforms": {
+          "type": "object",
+          "additionalProperties": { "$ref": "#/definitions/TransformNode" }
+        },
+
+        // A mapping containing all the primitive values in the pipeline.
+        "values": {
+          "type": "object",
+          "additionalProperties": { "$ref": "#/definitions/Value" }
+        },
+
+        // A flat list of windowing strategies. These may be referenced repeatedly.
+        "windowingStrategies": {
+          "type": "object",
+          "additionalProperties": { "$ref": "#/definitions/WindowingStrategy" }
+        },
+
+        "coders": {
+          "type": "object",
+          "additionalProperties": { "$ref": "#/definitions/Coder" }
+        }
+
+        // TODO: aggregators
+        // TODO: pipeline options
+        // TODO: pipeline-level display metadata
+        // TODO: SDK capabilities
+      }
+    },
+
+    // A user-defined function.
+    //
+    // TODO: the exact details of this will be fleshed out as the Fn API matures.
+    "UserFn": {
+      "type": "object",
+      "additionalProperties": false,
+      "properties": {
+
+        // A cross-language, stable, unique identifier for the function. It should indicate
+        // enough information that naive optimization (or language-to-language replacement)
+        // may be possible, since the runner will have no insight into the SDK-specific data.
+        "urn": { "type": "string" },
+
+        // A URL for the container image containing the SDK that understands how to run
+        // this user-definable function.
```

Comment:
Shouldn't this be required as well?

This might be better off as a reference to an environment and we add the environments to the top level Pipeline object